### PR TITLE
[improvement](bitmap) Use shared_ptr in BitmapValue to avoid deep copying

### PR DESCRIPTION
**Pull request by mrhhsg**
# Proposed changes

Currently bitmapvalue type is copied between columns, it cost a lot of memory. Use a shared ptr in bitmap value to avoid copy data.

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.36 seconds
 stream load tsv:          424 seconds loaded 74807831229 Bytes, about 168 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230426084646_clickbench_pr_135595.html

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

File: be/src/util/bitmap_value.h [hunk 1]
```diff
@@ -1876,14 +1912,27 @@ class BitmapValue {
         return result;
     }
 
+    void _prepare_bitmap_for_write() {
+        if (!_bitmap) {
+            _bitmap = std::make_shared<detail::Roaring64Map>();
+            return;
+        }
+
+        if (_bitmap.use_count() > 1) {
```
Comment (yiguolei) [inline be/src/util/bitmap_value.h:1921]: 感觉这里用use count 可能不太好，不如记录一个标记位，比如is_owner 这样的，然后再copy constructor 还有operator=的时候，都把is_owner 设置为false；prepare for write的时候，设置为true。

Comment (yiguolei) [reply be/src/util/bitmap_value.h:1921]: 如果is owner == true，就直接return了。 我觉得一个bool 的判断效率要比 shared ptr use count 高

**Comment by mrhhsg** [comment]
run buildall

File: be/src/util/bitmap_value.h [hunk 2]
```diff
@@ -1155,6 +1155,45 @@ class BitmapValue {
...
+        _is_shared = true;
+        // should also set other's state to shared, so that other bitmap value will
+        // create a new bitmap when it wants to modify it.
+        const_cast<BitmapValue&>(other)._is_shared = true;
+    }
+
+    BitmapValue(BitmapValue&& other) {
+        _type = other._type;
+        _sv = other._sv;
+        _bitmap = std::move(other._bitmap);
```
Comment (yiguolei) [inline be/src/util/bitmap_value.h:1171]: is shared = true or false?

File: be/src/util/bitmap_value.h [hunk 3]
```diff
@@ -1876,14 +1930,35 @@ class BitmapValue {
         return result;
     }
 
+    void _prepare_bitmap_for_write() {
+        if (!_bitmap) {
+            _bitmap = std::make_shared<detail::Roaring64Map>();
```
Comment (yiguolei) [inline be/src/util/bitmap_value.h:1935]: _is_shared = false

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.94 seconds
 stream load tsv:          423 seconds loaded 74807831229 Bytes, about 168 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.4 seconds inserted 10000000 Rows, about 125K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230522045521_clickbench_pr_148057.html

**Comment by mrhhsg** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.68 seconds
 stream load tsv:          433 seconds loaded 74807831229 Bytes, about 164 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          80.4 seconds inserted 10000000 Rows, about 124K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230522103952_clickbench_pr_148293.html

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**